polars/polars-core/src/chunked_array/ops/unique/rank.rs

```diff
@@ -57,6 +61,59 @@ pub(crate) fn rank(s: &Series, method: RankMethod) -> Series {
     use RankMethod::*;
     match method {
         Ordinal => inv_ca.into_series(),
+        Random => {
+            // Safety:
+            // in bounds
+            let arr = unsafe { s.take_unchecked(&sort_idx_ca).unwrap() };
+            let not_consecutive_same = (&arr.slice(1, len - 1))
+                .neq(&arr.slice(0, len - 1))
+                .rechunk();
+            let obs = not_consecutive_same.downcast_iter().next().unwrap();
+
+            // Collect slice indices for sort_idx which point to ties in the original series.
+            let mut ties_indices = Vec::with_capacity(len + 1);
+            let mut ties_index: usize = 0;
+
+            ties_indices.push(ties_index);
+            obs.iter().for_each(|b| {
+                if let Some(b) = b {
+                    ties_index += 1;
+                    if b {
+                        ties_indices.push(ties_index)
+                    }
+                }
+            });
+            // Close last slice (if there where nulls in the original series, they will always be in the last slice).
+            ties_indices.push(len);
+
+            let mut sort_idx = sort_idx.to_vec();
+
+            let rng = &mut thread_rng();
+
+            // Shuffle sort_idx positions which point to ties in the original series.
```

Comment:
@ritchie46: Can this be written as an iterator?

Reply:
I think so, but as you shuffle `sort_idx` which is outside of the loop, it doesn't really matter IMO.